### DISCUSSION TITLE
submitRecordToCommunities

### DISCUSSION THREAD

**Discussion by ThierryO** [eblondel/zen4R#158] [Q&A] [answered]
When should I run `submitRecordToCommunities()`? It seems logical to do it after `depositRecord()`. Does the record needs to be published too? Or can I run `submitRecordToCommunities()` on a draft version.

Is the `message` argument mandatory?

**Comment by eblondel** [comment]
This is real new, i just commited the function. I'm not sure this can be
done a draft, i didnt test this yet. I run it over a published record.
Message is optional.

Le ven. 24 mai 2024 à 21:19, Thierry Onkelinx ***@***.***> a
écrit :

**Reply by eblondel** [reply]
@ThierryO I had further look. 

The API method behind `submitRecordToCommunities()` assumes the record is published. If you try to submit a draft to a community, the method will fail with this message ` "The persistent identifier is not registered."`. So this method, for the timebeing, doesn't reproduce what we had in the past with the capacity to add a community to a draft.

In the UI, we can indeed select a community in a draft. By doing this, you can't publish by yourself the record, but a request is registered into the community. A request can be canceled by the submitter, and declined/accepted by the community admin/manager/curator. If the latter accept the record, the record will be published (so you give hand to this manager to do the actual record publication).

It's still not clear where we can specify the community in the draft model. I do not see it as part of the new record metadata model. I will further investigate if/how we can do it with the API.

**Reply by ThierryO** [reply]
Thanks for looking into this.

I'm updating [`checklist::bookdown_zenodo()`](https://inbo.github.io/checklist/reference/bookdown_zenodo.html). This function renders bookdown, creates the Zenodo records, updates it with the metadata from the bookdown, uploads the files and deposits everything as a draft publication. Then the user can do a manual publication.

We could either change it to do a full publication instead of a draft publication or add the communities in a separate step.

**Reply by eblondel** [reply]
I've added additional methods through https://github.com/eblondel/zen4R/issues/159 When in the UI you select a community on top of the record, we have now the equivalent with the method `createReviewRequest`. The terminology used by Zenodo is a bit confused, as in that case they indicate the following message:

![image](https://github.com/eblondel/zen4R/assets/1324396/54b0ee2d-737a-404c-9db8-b09c77e99322)

In reality, the purpose is to request to a community the record review. One limitation is that you can do it only for one `community`, not various, whereas for published records (with the method `submitRecordToCommunities`) we can effectively request the addition to more than one community.

After creating the review request, you can still delete it with `deleteReviewRequest`. To submit the record for review, you need to use `submitRecordForReview`, which goes in replacement of the `publishRecord` (publication without third-party review). As said in the previous comment, the publication will be then delegated to the community manager.

**Reply by ThierryO** [reply]
A follow-up question on the reviews. Creating and requesting a review works fine. Deleting the review returns a 400 status. 

```
manager <- ZenodoManager$new(sandbox = TRUE, token = sandbox_token)
x <- manager$depositRecord(zen_rec, publish = FALSE)
manager$deleteReviewRequest(x)
```

**Reply by eblondel** [reply]
`deleteReviewRequest` accepts the record itself (here `zen_rec`) as argument to delete. There can be only one review for a single record. This is also likely to be a limitation, you can ask the review to only one community

You can then do:
`manager$deleteReviewRequest(zen_rec)`

**Reply by ThierryO** [reply]
The `x` in my example is a `ZenodoRecord`. I created the review with `manager$createReviewRequest(record = x)` and submitted it with `manager$submitRecordForReview(recordId = x$id)`. I've added only one community. The owner of the sandbox account is also the maintainer of the community that I created for testing.

**Reply by eblondel** [reply, marked as answer]
oh sorry, I understood the issue:
* the `deleteReviewRequest` only applies to delete the selection of a community for a record; I can admit it is confusing but that's how the API seems to be designed. 
* Once you had submitted the record for review, you can't delete the selection of the community (what is done with `deleteReviewRequest`), but rather you have to cancel the review. For this you need to retrieve the review request

Full example:

```R
my_community = "openfair"
record_id = 99999999
record = ZENODO$getDepositionById(record_id)
ZENODO$createReviewRequest(record, my_community) #should be ok
ZENODO$submitRecordForReview(record$id) #should be ok
ZENODO$deleteReviewRequest(record) #can't work because you submitted the record for review

#to cancel the the submission for record to review, use the requests API
#fetch first the community
zen_com = ZENODO$getCommunityById(my_community)
#retrieve the request that was created with the review submission (using status the target community and the target record)
reqs = ZENODO$getRequests(q = sprintf("status:submitted AND receiver.community:%s AND topic.record:%s", zen_com$id, record$id))
ZENODO$cancelRequest(reqs[[1]]$id) #return TRUE if canceled
```
